### PR TITLE
Datasource authentication updates

### PR DESCRIPTION
**Pull request by khansaad**
## Description

This PR contains new feature addition of supporting different REST API authentication for Datasource, DB, cloudwatch and other services. 

Fixes # (issue)

### Type of change

- [ ] Bug fix
- [x] New feature
- [ ] Docs update
- [ ] Breaking change (What changes might users need to make in their application due to this PR?)
- [ ] Requires DB changes

## How has this been tested?
Testing is in progress

Please describe the tests that were run to verify your changes and steps to reproduce. Please specify any test configuration required. 

- [ ] New Test X
- [ ] Functional testsuite

**Test Configuration**
* Kubernetes clusters tested on: Resource Hub cluster

## Checklist :dart:

- [x] Followed coding guidelines
- [x] Comments added
- [x] Dependent changes merged
- [ ] Documentation updated
- [ ] Tests added or updated

## Additional information
Tested using this custom image: `quay.io/khansaad/autotune_operator:ds-auth`

Screenshot from 2024-09-12 21-17-40

Shows the failure when wrong token path is passed :

Screenshot from 2024-09-12 22-43-21

### CODE REVIEW

File: src/main/java/com/autotune/common/auth/AuthenticationConfig.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    @Override
+    public String toString() {
+        return "AuthenticationConfig{" +
+                "type='" + type + '\'' +
+                ", credentials=" + credentials +
+                '}';
+    }
+}
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/common/auth/AuthenticationConfig.java:57]: EOF missing

File: src/main/java/com/autotune/operator/KruizeDeploymentInfo.java [hunk 2]
```diff
@@ -38,8 +38,9 @@
  */
 public class KruizeDeploymentInfo {
     private static final Logger LOGGER = LoggerFactory.getLogger(KruizeDeploymentInfo.class);
-    public static String database_username;
-    public static String database_password;
+    public static String database_authentication_type;
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/operator/KruizeDeploymentInfo.java:41]: why this change required ? 
this change will affect ROS

File: manifests/crc/default-db-included-installation/openshift/kruize-crc-openshift.yaml [hunk 3]
```diff
@@ -75,10 +75,15 @@ data:
         "adminUsername": "admin",
         "hostname": "kruize-db-service",
```
Comment (msvinaykumar) [inline manifests/crc/default-db-included-installation/openshift/kruize-crc-openshift.yaml:76]: I guess this PR is limited to datasource auth , But not for DB ?

File: manifests/crc/default-db-included-installation/openshift/kruize-crc-openshift.yaml [hunk 4]
```diff
@@ -92,7 +97,7 @@ data:
       "savetodb": "true",
       "dbdriver": "jdbc:postgresql://",
       "plots": "true",
-      "local": "false",
+      "local": "true",
```
Comment (msvinaykumar) [inline manifests/crc/default-db-included-installation/openshift/kruize-crc-openshift.yaml:100]: let default value set to false , Until ROS move into thanos

File: manifests/crc/default-db-included-installation/openshift/kruize-crc-openshift.yaml [hunk 5]
```diff
@@ -107,12 +112,17 @@ data:
       },
       "logging" : {
         "cloudwatch": {
-          "accessKeyId": "",
           "logGroup": "kruize-logs",
           "logStream": "kruize-stream",
           "region": "",
-          "secretAccessKey": "",
           "logLevel": "INFO"
+          "authentication": {
```
Comment (msvinaykumar) [inline manifests/crc/default-db-included-installation/openshift/kruize-crc-openshift.yaml:119]: Let make this PR limited scope .. Only datasource authentication

File: src/main/java/com/autotune/common/auth/Credentials.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+                ", clientSecret='" + clientSecret + '\'' +
+                ", username='" + username + '\'' +
+                ", password='" + password + '\'' +
+                ", tokenEndpoint='" + tokenEndpoint + '\'' +
+                ", tokenFilePath='" + tokenFilePath + '\'' +
+                ", apiKey='" + apiKey + '\'' +
+                ", headerName='" + headerName + '\'' +
+                '}';
+    }
+}
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/common/auth/Credentials.java:108]: EOF missing

File: src/main/java/com/autotune/common/datasource/prometheus/PrometheusDataOperatorImpl.java [hunk 7]
```diff
@@ -127,18 +134,21 @@ public Object getValueForQuery(String url, String query) {
...
      */
     @Override
-    public JSONObject getJsonObjectForQuery(String url, String query) {
-        GenericRestApiClient apiClient = new GenericRestApiClient(
-                CommonUtils.getBaseDataSourceUrl(
-                        url,
-                        KruizeConstants.SupportedDatasources.PROMETHEUS
-                )
-        );
+    public JSONObject getJsonObjectForQuery(String url, String query, AuthenticationConfig authenticationConfig) {
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/common/datasource/prometheus/PrometheusDataOperatorImpl.java:144]: notice that many functions are using authenticationConfig. This configuration should be set at the experiment level, similar to how experiment.setDatasource works. When retrieving the datasource using experiment.getDatasource, it should also return the associated authenticationConfig.

To achieve this, the authenticationConfig object should be linked to the datasource and accessible through it.

File: src/main/java/com/autotune/utils/CloudWatchAppender.java [hunk 8]
```diff
@@ -73,7 +73,7 @@ public void append(LogEvent event) {
     }
 
     public static void configureLoggerForCloudWatchLog() {
-        if (cloudwatch_logs_access_key_id != null && !cloudwatch_logs_access_key_id.isEmpty() && cloudwatch_logs_secret_access_key != null && !cloudwatch_logs_secret_access_key.isEmpty() && cloudwatch_logs_region != null && !cloudwatch_logs_region.isEmpty()) {
+        if (cloudwatch_logs_authentication_credentials_access_key_id != null && !cloudwatch_logs_authentication_credentials_access_key_id.isEmpty() && cloudwatch_logs_authentication_credentials_secret_access_key != null && !cloudwatch_logs_authentication_credentials_secret_access_key.isEmpty() && cloudwatch_logs_region != null && !cloudwatch_logs_region.isEmpty()) {
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/utils/CloudWatchAppender.java:76]: lets limit this change only related to only datasource

File: src/main/java/com/autotune/experimentManager/handler/MetricCollectionHandler.java [hunk 9]
```diff
@@ -160,9 +162,10 @@ public void execute(ExperimentTrial experimentTrial, TrialDetails trialDetails,
                                 }
                                 if (null != updatedContainerQuery) {
                                     LOGGER.debug("Updated Query - " + updatedContainerQuery);
+                                    AuthenticationConfig noAuthConfig = AuthenticationConfig.noAuth();
                                     String queryResult = (String) ado.getValueForQuery(experimentTrial.getDatasourceInfoHashMap()
                                             .get(containerMetric.getDatasource())
-                                            .getUrl().toString(), updatedContainerQuery);
+                                            .getUrl().toString(), updatedContainerQuery, noAuthConfig);
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/experimentManager/handler/MetricCollectionHandler.java:168]: authoconfig should get tagged to datasource obj and 
datasource obj should get tagged to Experiment

so data pass using parameter should be just experiment Obj everywhere

File: src/main/java/com/autotune/operator/KruizeDeploymentInfo.java (hunk 2 above)
Comment (khansaad) [reply src/main/java/com/autotune/operator/KruizeDeploymentInfo.java]: Stashed for now. Will be updating later on for common auth feature

File: manifests/crc/default-db-included-installation/openshift/kruize-crc-openshift.yaml (hunk 3 above)
Comment (khansaad) [reply manifests/crc/default-db-included-installation/openshift/kruize-crc-openshift.yaml]: Reverted

**Review by msvinaykumar — APPROVED** [review]
LGTM

Review Checklist

#### Common Review
- [x] Code formatting adheres to the project’s style guide.
- [x] No unnecessary commented-out code.
- [x] Appropriate use of comments and documentation.

#### Logical Flow Review
- [ ] Code logic is clear and easy to follow.
- [ ] All branches and conditions are covered and tested.
- [ ] No redundant or duplicate code.

#### Demo Update
- [ ] Demonstration examples are updated to reflect changes.
- [ ] README and documentation include updated demo instructions.
- [ ] Any necessary demo files are included and tested.

#### Naming Conventions
- [ ] Variable and function names follow the project's naming conventions.
- [ ] File names are descriptive and adhere to naming standards.
- [ ] No ambiguous or misleading names.

#### Workflow
- [ ] CI/CD pipelines are updated and passing.
- [ ] Tests are included and all pass.
- [ ] The workflow for deploying or integrating changes is clear and documented.

#### Does It Affect ROS 
- [ ] Changes are reviewed for compatibility with ROS.
- [ ] ROS-specific configurations are updated if necessary.
- [ ] Any ROS dependencies or integrations are tested and verified.

**Comment by chandrams** [comment]
@khansaad - Can you resolve the conflicts.

**Comment by khansaad** [comment]
> @khansaad - Can you resolve the conflicts.

Done

**Review by chandrams — APPROVED** [review]
LGTM

**Comment by chandrams** [comment]
@khansaad - I see a merge conflict, can you please resolve this. Thanks.

File: manifests/crc/BYODB-installation/openshift/kruize-crc-openshift.yaml [hunk 10]
```diff
@@ -68,7 +68,22 @@ data:
...
-      }
+      },
+      "datasource": [
+        {
+          "name": "prometheus-1",
+          "provider": "prometheus",
+          "serviceName": "prometheus-k8s",
+          "namespace": "openshift-monitoring",
+          "url": "",
+          "authentication": {
```
Comment (msvinaykumar) [inline manifests/crc/BYODB-installation/openshift/kruize-crc-openshift.yaml:79]: Ensure that this change, including the check for the authentication object, is applied only when local=true. The Kruize pods should not fail if the authentication object is absent, as sometimes datasource authentication is not required.

File: src/main/java/com/autotune/common/auth/AuthenticationStrategyFactory.java [hunk 11]
```diff
@@ -0,0 +1,29 @@
+package com.autotune.common.auth;
+
+import com.autotune.utils.KruizeConstants;
+
+public class AuthenticationStrategyFactory {
+
+    public static AuthenticationStrategy createAuthenticationStrategy(AuthenticationConfig authConfig) {
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/common/auth/AuthenticationStrategyFactory.java:7]: Also support for no auth stratgey

File: manifests/crc/BYODB-installation/openshift/kruize-crc-openshift.yaml (hunk 10 above)
Comment (khansaad) [reply manifests/crc/BYODB-installation/openshift/kruize-crc-openshift.yaml:79]: The datasource object is applicable only for local monitoring. Therefore, any code changes related to it have been made specifically within the local monitoring context and do not affect the ROS case.

File: src/main/java/com/autotune/common/auth/AuthenticationStrategyFactory.java (hunk 11 above)
Comment (khansaad) [reply src/main/java/com/autotune/common/auth/AuthenticationStrategyFactory.java:7]: Support for no auth has been added in the AuthenticationConfig class:

https://github.com/khansaad/autotune/blob/f0c5b16d9759c62a04feff09597a7d9ce99e4ec7/src/main/java/com/autotune/common/auth/AuthenticationConfig.java#L64
